<a href="https://colab.research.google.com/github/IyadSultan/educational/blob/main/training_a_transformer_NER.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Training a Transformer for Medical Problem NER in Clinical Text

**Named Entity Recognition (NER)** in healthcare can automatically highlight medical problems (like diseases or symptoms) in clinical notes, helping clinicians quickly identify key conditions. In this hands-on tutorial, we will fine-tune a Transformer model to recognize medical problems in text using Hugging Face 🤗 Transformers.

We'll use a domain-specific BERT model (BioClinicalBERT) and an open medical NER dataset, walking through the entire process from data loading to evaluation. What you will learn:

1- Setting up the environment and installing necessary libraries.

2- Loading an open-access medical NER dataset from the Hugging Face Hub.

3- Preprocessing clinical text data and aligning entity labels with subword tokens.

4- Choosing a suitable pre-trained Transformer model for clinical text.

5- Fine-tuning the model for NER using the Hugging Face Trainer API.

6- Evaluating model performance (precision, recall, F1-score).

7- Testing the model on example clinical text.


*Prerequisites:* Basic Python coding and familiarity with general concepts of machine learning. We will explain NLP and Transformer concepts in simple terms, so healthcare professionals with beginner coding experience should be able to follow along.

# 1. Environment Setup
First, we need to set up our environment. This tutorial is designed for Google Colab – make sure you've selected a GPU runtime for faster training (go to Runtime > Change runtime type > Hardware accelerator > GPU). We’ll install Hugging Face’s Transformers, Datasets, and evaluation libraries.

In [ ]:
!pip install transformers datasets evaluate seqeval

*Explanation* : The transformers library provides the model and training tools, datasets will let us easily load the dataset, and seqeval (via evaluate) is a library for computing NER metrics like precision and F1. After running this, you should see the libraries installing.

In [42]:
# Let’s also verify that a GPU is available:
import torch
print("GPU available:", torch.cuda.is_available())


GPU available: True


If the output says GPU available: True, we’re ready to go (Colab Pro’s GPU will help speed up training). If it’s False, double-check the runtime settings.

# 2. Loading the Dataset
We will use an open-access dataset from the Hugging Face Hub. One good option is the NCBI Disease corpus, which contains biomedical text (PubMed abstracts) annotated with disease names​. This **English-language dataset** is open-access and suitable for our task of extracting medical problems. Let's load the dataset using 🤗 Datasets:

In [43]:
from datasets import load_dataset

# Load the NCBI Disease dataset (with train/val/test splits)
raw_datasets = load_dataset("ncbi_disease")
print(raw_datasets)


DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'ner_tags'],
        num_rows: 5433
    })
    validation: Dataset({
        features: ['id', 'tokens', 'ner_tags'],
        num_rows: 924
    })
    test: Dataset({
        features: ['id', 'tokens', 'ner_tags'],
        num_rows: 941
    })
})


This should download the dataset and show the splits (train/validation/test). The NCBI Disease corpus has 793 abstracts with disease annotations​, already split into about 5,433 training sentences, 924 validation sentences, and 941 test sentences. Each data sample in this dataset is a sentence with two fields:

- "tokens": the list of word tokens in the sentence.
- "ner_tags": the list of numeric labels for each token (0 = outside any entity, 1 = beginning of a disease entity, 2 = inside a disease entity)​.

Let’s inspect an example from the training set:

In [44]:
# Peek at one training example
example = raw_datasets["train"][0]
print(example["tokens"])
print(example["ner_tags"])


['Identification', 'of', 'APC2', ',', 'a', 'homologue', 'of', 'the', 'adenomatous', 'polyposis', 'coli', 'tumour', 'suppressor', '.']
[0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 2, 2, 0, 0]


In this example, the tokens ['adenomatous', 'polyposis', 'coli', 'tumour'] have tags [1, 2, 2, 2], meaning "adenomatous" is tagged as B-disease (beginning of a disease name) and the next three tokens are I-disease (continuation of the disease name), while all other tokens are 0 (not an entity). The disease mentioned here is “adenomatous polyposis coli tumour.”

**Understanding the task**: Our goal is to train a model that takes a sequence of tokens (a clinical sentence) and outputs the correct tag (O, B-problem, I-problem) for each token, thereby identifying spans of text that are medical problems (diseases, in this dataset).

# 3. Data Preprocessing and Tokenization
Before training, we need to preprocess the data for our Transformer model. Transformers like BERT cannot take raw text strings directly for NER; we must:
Tokenize the text with the model’s tokenizer, and Align the labels to the tokenizer’s output tokens.

*Why align labels?*

Models like BERT use subword tokenization. A single word (e.g., "diabetes") might be broken into multiple subtokens (e.g., "dia", "##betes"). Our dataset labels are at the word level, so we need to create a label for each subtoken as well. Typically, we give the same label to all subtokens of a word, or label only the first subtoken and mark the rest as “ignore” in the loss calculation. We will use the Hugging Face AutoTokenizer to handle tokenization. Let’s choose our model’s tokenizer (we will use BioClinicalBERT in the next section, which has a fast tokenizer available):

In [45]:
from transformers import AutoTokenizer

model_checkpoint = "emilyalsentzer/Bio_ClinicalBERT"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)


Now we define a function to tokenize the inputs and align the NER tags. We'll use is_split_into_words=True because our dataset already provides a list of tokens for each example (so we don't want the tokenizer to re-split the sentence into words, but rather tokenize each given token).

**For each tokenized example:**
- We get a mapping of token indices to original word indices using tokenized_inputs.word_ids(batch_index).
= We assign each subtoken the label of its originating word. We will mark only the first subtoken of each word with the original label and use -100 for the remaining subtokens. The value -100 is a special label that tells the loss function to ignore those positions (so we don't double-count a multi-subtoken word).

In [46]:
def tokenize_and_align_labels(examples):
    # Use padding and truncation to ensure consistent sequence lengths
    tokenized_inputs = tokenizer(
        examples["tokens"],
        is_split_into_words=True,
        truncation=True,
        padding="max_length",
        max_length=128  # Choose an appropriate max length
    )

    all_labels = examples["ner_tags"]
    aligned_labels = []

    for i, labels in enumerate(all_labels):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        label_ids = []
        previous_word_idx = None

        for word_idx in word_ids:
            if word_idx is None:
                # Special token like [CLS] or [SEP]
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                # Start of a new word, use that word's label
                label_ids.append(labels[word_idx])
            else:
                # Continuation of the same word, mark for ignoring
                label_ids.append(-100)
            previous_word_idx = word_idx

        aligned_labels.append(label_ids)

    # Include the aligned labels in the tokenized input
    tokenized_inputs["labels"] = aligned_labels
    return tokenized_inputs

# Apply the tokenization and alignment to the entire dataset
tokenized_datasets = raw_datasets.map(tokenize_and_align_labels, batched=True)

Map:   0%|          | 0/5433 [00:00<?, ? examples/s]

Map:   0%|          | 0/924 [00:00<?, ? examples/s]

Map:   0%|          | 0/941 [00:00<?, ? examples/s]

*Explanation:* We iterate through each example’s word indices from the tokenizer. When word_idx is None, it corresponds to special tokens (like [CLS] start or [SEP] end tokens) – we assign -100 so that they are ignored in training. When we see a new word (word index not equal to the previous one), we take that word’s original label (labels[word_idx]). If the tokenizer output is still on the same word as the previous token (meaning the word was split into multiple subtokens), we also assign -100 to those subtokens. This way, only the first subtoken of a word carries the label, and the loss will be calculated only once per original word.

**Note: An alternative would be to label subsequent subtokens as "I-" (inside) tags of the same entity. In our simple approach, we ignore them for loss calculation to avoid overweighting long words. This is a common technique to handle subword tokenization in NER​**.

Now that we have tokenized_datasets, each entry has not only "input_ids" and "attention_mask" (from tokenization) but also an aligned "labels" sequence for training. We can verify the tokenization on our earlier example to see how the labels align:

In [68]:
# Check tokenization and label alignment on the example
tokens = tokenizer(example["tokens"], is_split_into_words=True)
print("Original tokens:", example["tokens"])
print("Original labels:", example["ner_tags"])
print("Subword tokens:", tokens.tokens())
print("Aligned labels:", tokenize_and_align_labels({"tokens": [example["tokens"]], "ner_tags": [example["ner_tags"]]})["labels"][0])


Original tokens: ['The', 'risk', 'of', 'cancer', ',', 'especially', 'lymphoid', 'neoplasias', ',', 'is', 'substantially', 'elevated', 'in', 'A', '-', 'T', 'patients', 'and', 'has', 'long', 'been', 'associated', 'with', 'chromosomal', 'instability', '.']
Original labels: [0, 0, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 1, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Subword tokens: ['[CLS]', 'the', 'risk', 'of', 'cancer', ',', 'especially', 'l', '##ymph', '##oid', 'neo', '##p', '##lasia', '##s', ',', 'is', 'substantially', 'elevated', 'in', 'a', '-', 't', 'patients', 'and', 'has', 'long', 'been', 'associated', 'with', 'ch', '##rom', '##oso', '##mal', 'instability', '.', '[SEP]']
Aligned labels: [-100, 0, 0, 0, 1, 0, 0, 1, -100, -100, 2, -100, -100, -100, 0, 0, 0, 0, 0, 1, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, -100, -100, -100, 0, 0, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -1

This will show the original tokens and their labels, the subword tokens (including special [CLS] and [SEP]), and the aligned labels for each subword token. You should see -100 for the special tokens and second part of any split word. For instance, if "polyposis" was split into ["poly", "##posis"], the first subtoken might get label 2 (I-disease) and the second subtoken -100.

Let's also visualize an example from the training set in a more readable format:

In [ ]:
from datasets import load_dataset
from transformers import AutoTokenizer
import pandas as pd
from IPython.display import display

# Define the label mapping
id2label = {0: "O", 1: "B-Disease", 2: "I-Disease"}

# 1. Display a raw example from the training set
example_idx = 10  # You can change this index to view different examples
example = raw_datasets["train"][example_idx]

print("=== RAW TRAINING EXAMPLE ===")
df_raw = pd.DataFrame({
    "Token": example["tokens"],
    "NER Tag ID": example["ner_tags"],
    "NER Tag": [id2label[tag] for tag in example["ner_tags"]]
})
display(df_raw)

# Extract entities for easier viewing
entities = []
current_entity = []
for token, tag_id in zip(example["tokens"], example["ner_tags"]):
    if tag_id == 1:  # B-Disease
        if current_entity:
            entities.append(" ".join(current_entity))
            current_entity = []
        current_entity.append(token)
    elif tag_id == 2:  # I-Disease
        current_entity.append(token)
    else:  # O
        if current_entity:
            entities.append(" ".join(current_entity))
            current_entity = []

if current_entity:
    entities.append(" ".join(current_entity))

print("\nEntities found:")
for i, entity in enumerate(entities):
    print(f"{i+1}. {entity}")

print("\nOriginal sentence:")
print(" ".join(example["tokens"]))

# Display the tokenized example
print("\n=== TOKENIZED EXAMPLE ===")
# Show how tokens are aligned with labels after tokenization
tokenized_inputs = tokenizer(
    example["tokens"],
    is_split_into_words=True,
    truncation=True,
    padding="max_length",
    max_length=128
)

# Manual label alignment for display
word_ids = tokenized_inputs.word_ids()
labels = example["ner_tags"]
aligned_labels = []
previous_word_idx = None

for word_idx in word_ids:
    if word_idx is None:
        # Special token
        aligned_labels.append(-100)
    elif word_idx != previous_word_idx:
        # Start of a new word
        aligned_labels.append(labels[word_idx])
    else:
        # Continuation of same word
        aligned_labels.append(-100)
    previous_word_idx = word_idx

# Display in DataFrame for clarity
tokens = tokenizer.convert_ids_to_tokens(tokenized_inputs["input_ids"])
df_tokenized = pd.DataFrame({
    "Index": range(len(tokens)),
    "Token": tokens,
    "Word ID": word_ids,
    "Label ID": aligned_labels,
    "Label": [id2label[l] if l != -100 else "IGNORE" for l in aligned_labels]
})
display(df_tokenized)

# Show some statistics
special_tokens = sum(1 for l in aligned_labels if l == -100)
real_labels = sum(1 for l in aligned_labels if l != -100)

print(f"\nTotal tokens after tokenization: {len(tokens)}")
print(f"Special tokens or subword continuations (IGNORE): {special_tokens}")
print(f"Tokens with real labels: {real_labels}")
print(f"Original tokens: {len(example['tokens'])}")

# 4. Choosing a Pretrained Model (BioClinicalBERT)
Selecting a good pretrained model as a starting point is crucial. We need a model that understands clinical language. BioClinicalBERT is an excellent choice: it’s a BERT-based model that was pretrained on biomedical literature and clinical notes (MIMIC-III ICU records), giving it a strong grasp of medical terminology and context. This domain-specific pretraining should help it recognize disease names and other medical entities better than a generic BERT.

Other options could be BioBERT (pretrained on biomedical articles) or even a smaller model like DistilBERT fine-tuned on a medical NER dataset. However, BioClinicalBERT is well-suited for clinical text, and since we have a GPU, its size (base BERT, ~110M parameters) is manageable.

Let's load the model with a classification head for token classification (NER). We will use AutoModelForTokenClassification which adds a token-level classification layer on top of the transformer. We need to specify the number of labels our model will predict:

- Label 0: "O" (no entity)
- Label 1: "B-Disease" (beginning of a disease mention)
- Label 2: "I-Disease" (inside a disease mention)

We'll also pass **id2label and label2id** mappings for better clarity (so the model knows which label index corresponds to "B-Disease", etc., which will be useful for inference).

In [69]:
from transformers import AutoModelForTokenClassification
import numpy as np
from sklearn.utils.class_weight import compute_class_weight

# Extract all labels from training data to compute class weights
all_labels = []
for example in tokenized_datasets["train"]:
    all_labels.extend([l for l in example["labels"] if l != -100])

# Compute class weights to handle imbalanced data
class_weights = compute_class_weight(
    class_weight='balanced',
    classes=np.array([0, 1, 2]),  # O, B-Disease, I-Disease
    y=np.array(all_labels)
)

print(f"Computed class weights: {class_weights}")

# Define our label mappings
num_labels = 3  # O, B-Disease, I-Disease
id2label = {0: "O", 1: "B-Disease", 2: "I-Disease"}
label2id = {"O": 0, "B-Disease": 1, "I-Disease": 2}

# Create the model
model = AutoModelForTokenClassification.from_pretrained(
    model_checkpoint,
    num_labels=num_labels,
    id2label=id2label,
    label2id=label2id
)

Computed class weights: [0.36342693 8.81511733 7.40884821]


When this runs, it will download the BioClinicalBERT weights. You might see a warning that some weights are not used (those are for the original masked language modeling head) and some new weights are initialized (the classification head). That’s expected because we are repurposing the base model for NER.

**Note: BioClinicalBERT was initialized from BioBERT and trained on. It’s a domain-specific model ideal for clinical text. If you were dealing with general text, you might use a generic BERT base model, but for our medical task, leveraging domain knowledge should give us a boost.**

# 5. Fine-Tuning the Model with Trainer API
Now we have our prepared dataset (tokenized_datasets) and our model. The next step is to fine-tune the model on the training data. We'll use Hugging Face’s Trainer API, which simplifies the training loop and handles things like gradient accumulation, evaluation, and more. Setup Training Arguments: We need to specify how we train, e.g., number of epochs, batch size, learning rate, etc. For NER with a base BERT model:
- A few epochs (2-4) are usually sufficient for a dataset of this size.
- A learning rate around 2e-5 to 5e-5 works well for fine-tuning BERT.
- We’ll use the validation set to evaluate after each epoch.

We also define a compute_metrics function to compute precision, recall, and F1 using the seqeval metric. We will focus on the “Disease” entity class performance and overall metrics.

In [70]:
import numpy as np
import evaluate

# Load the seqeval metric for NER evaluation
seqeval = evaluate.load("seqeval")

# Define compute_metrics to use during evaluation
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    true_labels = []
    true_predictions = []
    for pred, lab in zip(predictions, labels):
        # Remove ignored index (subword pieces)
        lab = [l for l in lab if l != -100]
        pred = pred[:len(lab)]  # truncate prediction to same length as lab
        true_labels.append([id2label[l] for l in lab])
        true_predictions.append([id2label[p] for p in pred])
    # Compute metrics using seqeval
    results = seqeval.compute(predictions=true_predictions, references=true_labels)
    # Return precision, recall, F1, and accuracy
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

*A quick explanation:* We take the logits (model outputs before softmax) and labels from the evaluation batch. We argmax the logits to get the predicted label indices. We then filter out any -100 in the labels (those correspond to subword pieces we chose to ignore), and make sure to truncate the predictions to the same length (since we don't want to count predictions for subword tokens that were ignored in labels). We map the numeric labels back to their string names (using id2label) and feed the lists to seqeval.compute(). Seqeval will calculate precision, recall, F1, etc., treating the labels as sequences of entity tags. We return the overall metrics. (Seqeval also provides class-specific metrics, but “overall” here mostly reflects the single entity type we care about, Disease.)

In [71]:
import torch
from transformers import Trainer, TrainingArguments

# Create a custom trainer with weighted loss to handle class imbalance
class WeightedLossTrainer(Trainer):
    def __init__(self, class_weights=None, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # Convert to the same data type as the model (float32)
        if class_weights is not None:
            self.class_weights = torch.tensor(class_weights, dtype=torch.float32)
        else:
            self.class_weights = None

    def compute_loss(self, model, inputs, return_outputs=False, **kwargs):
        labels = inputs.pop("labels")
        outputs = model(**inputs)
        logits = outputs.logits

        # Use class weights in loss calculation
        if self.class_weights is not None:
            loss_fct = torch.nn.CrossEntropyLoss(
                weight=self.class_weights.to(model.device),
                ignore_index=-100
            )
        else:
            loss_fct = torch.nn.CrossEntropyLoss(ignore_index=-100)

        loss = loss_fct(logits.view(-1, model.config.num_labels), labels.view(-1))

        return (loss, outputs) if return_outputs else loss

# Set up training arguments
training_args = TrainingArguments(
    output_dir="./output_model_dir",
    num_train_epochs=8,
    per_device_train_batch_size=16,
    learning_rate=1e-4,  # Higher learning rate for faster convergence
    weight_decay=0.01,   # Regularization to prevent overfitting
    logging_steps=len(tokenized_datasets["train"]) // 16,  # Log once per epoch
    fp16=True            # Mixed precision training for better performance
)

# Create the weighted loss trainer
trainer = WeightedLossTrainer(
    class_weights=class_weights,
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

<ipython-input-71-a758a5c44360>:7: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `WeightedLossTrainer.__init__`. Use `processing_class` instead.
  super().__init__(*args, **kwargs)


Key points in these settings:
We set evaluation_strategy="epoch" to evaluate on the validation set at the end of each epoch.
We chose 3 epochs. You can adjust this; often 2-4 is enough for convergence on a small dataset, and more can cause overfitting.
per_device_train_batch_size=16 is reasonable for a base model on a GPU with enough memory (Colab Pro GPUs can handle this). If you get OOM errors, you might reduce the batch size.
We disable saving checkpoints (save_strategy="no") and pushing to hub for simplicity, but in a real scenario, you might want to save the fine-tuned model.
Before training, let's optionally evaluate the untrained model on the validation set to get a baseline. (Since the classification head is random initially, we expect it to perform very poorly – likely predicting no entities correctly.)

In [72]:
# Before training, let's evaluate the untrained model to get a baseline:
# Evaluate the model *before* training (baseline performance)
baseline_metrics = trainer.evaluate()
print("Baseline performance:", baseline_metrics)


{'eval_loss': 1.0683281421661377, 'eval_model_preparation_time': 0.0032, 'eval_precision': 0.0116093785567949, 'eval_recall': 0.06480304955527319, 'eval_f1': 0.01969111969111969, 'eval_accuracy': 0.6922274604697735, 'eval_runtime': 3.0065, 'eval_samples_per_second': 307.335, 'eval_steps_per_second': 38.583}
Baseline performance: {'eval_loss': 1.0683281421661377, 'eval_model_preparation_time': 0.0032, 'eval_precision': 0.0116093785567949, 'eval_recall': 0.06480304955527319, 'eval_f1': 0.01969111969111969, 'eval_accuracy': 0.6922274604697735, 'eval_runtime': 3.0065, 'eval_samples_per_second': 307.335, 'eval_steps_per_second': 38.583}


This will run the model on the validation set and print metrics. Likely, you will see precision, recall, F1 near 0. The model probably labels everything as "O" (no entity) by default, resulting in 0 recall for the disease class. The accuracy might be high (because most tokens are non-entity and those were "predicted" correctly by always outputting O), but accuracy is misleading here since our interest is in the rare entity class. This highlights why we use precision/recall/F1 for NER evaluation – a model that just predicts no entities will have high token-level accuracy but F1 of 0 for the actual task. Now, let's fine-tune the model on the training data:

In [73]:
# Train the model
train_results = trainer.train()


{'loss': 0.2119, 'grad_norm': 10.693309783935547, 'learning_rate': 8.764705882352942e-05, 'epoch': 0.9970588235294118}
{'loss': 0.0817, 'grad_norm': 0.4537445306777954, 'learning_rate': 7.518382352941177e-05, 'epoch': 1.9941176470588236}
{'loss': 0.044, 'grad_norm': 9.398859977722168, 'learning_rate': 6.275735294117647e-05, 'epoch': 2.9911764705882353}
{'loss': 0.0216, 'grad_norm': 0.3768038749694824, 'learning_rate': 5.0294117647058826e-05, 'epoch': 3.988235294117647}
{'loss': 0.0141, 'grad_norm': 0.07289780676364899, 'learning_rate': 3.7830882352941175e-05, 'epoch': 4.985294117647059}
{'loss': 0.0088, 'grad_norm': 13.817891120910645, 'learning_rate': 2.536764705882353e-05, 'epoch': 5.982352941176471}
{'loss': 0.0038, 'grad_norm': 0.37299567461013794, 'learning_rate': 1.2904411764705885e-05, 'epoch': 6.979411764705882}
{'loss': 0.0022, 'grad_norm': 0.5072056651115417, 'learning_rate': 4.411764705882353e-07, 'epoch': 7.976470588235294}
{'train_runtime': 423.4081, 'train_samples_per_sec


These metrics provide valuable insights into how our model is learning:

1. **Loss**: We can see a dramatic decrease in training loss from 0.2119 in the first epoch to 0.0022 by the end of training - a 99% reduction. This indicates that our model is effectively learning to identify disease entities in the text.

2. **Learning Rate**: We're using a linear learning rate scheduler that gradually decreases from 8.76e-05 to nearly zero by the end of training. This helps the model converge to an optimal solution by taking smaller steps as it gets closer to the minimum.

3. **Gradient Norm**: The gradient norm fluctuates throughout training, with some spikes (like in epochs 1 and 6). These spikes represent moments when the model encounters batches of data that cause larger gradient updates. The overall trend shows the model is stabilizing, especially in the later epochs.

4. **Training Efficiency**: The final line shows our training took about 423 seconds (7 minutes) to complete all 8 epochs, processing approximately 102 examples per second. This is quite efficient for a GPU-based training run.

5. **Convergence Pattern**: The most important observation is how quickly the loss decreases in early epochs (from 0.21 to 0.08 in just one epoch) and then continues to improve more gradually. This is a healthy learning curve that suggests:
   - Our learning rate was well-chosen
   - The pre-trained BioClinicalBERT model provides an excellent starting point
   - The class weighting is effectively handling the imbalanced dataset

The final training loss of 0.048 averaged across all epochs (with the last epoch having a loss of just 0.0022) indicates that our model has successfully learned to identify disease entities in medical text.

*This pattern of rapidly decreasing loss followed by more gradual improvements is characteristic of transfer learning with pre-trained models - we're seeing the benefit of starting with a model that already understands biomedical language, then fine-tuning it for our specific NER task.*



This will start the training loop. You should see output for each epoch, including the evaluation metrics on the validation set at the end of each epoch.

For example, after each epoch you might see lines showing the eval precision, recall, f1, etc., gradually improving. Training may take a few minutes per epoch on Colab GPU. Once training is complete, we can evaluate the model on the test set (which the model has never seen, to get an unbiased performance estimate):

In [74]:
# Fix the compute_metrics function to properly handle token-level predictions
def compute_metrics_fixed(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)

    true_labels = []
    true_predictions = []

    for pred, lab in zip(predictions, labels):
        # Filter out ignored indexes (-100)
        valid_indices = [i for i, l in enumerate(lab) if l != -100]
        pred_filtered = [pred[i] for i in valid_indices]
        lab_filtered = [lab[i] for i in valid_indices]

        # Convert to string labels
        true_labels.append([id2label[l] for l in lab_filtered])
        true_predictions.append([id2label[p] for p in pred_filtered])

    # Compute metrics
    results = seqeval.compute(predictions=true_predictions, references=true_labels)

    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

# Re-evaluate with the fixed function
trainer.compute_metrics = compute_metrics_fixed
test_results = trainer.evaluate(tokenized_datasets["test"])
print("Fixed test results:", test_results)

{'eval_loss': 0.29126086831092834, 'eval_model_preparation_time': 0.0032, 'eval_precision': 0.825381679389313, 'eval_recall': 0.9010416666666666, 'eval_f1': 0.8615537848605578, 'eval_accuracy': 0.9841509742249092, 'eval_runtime': 2.3923, 'eval_samples_per_second': 393.338, 'eval_steps_per_second': 49.324, 'epoch': 8.0}
Fixed test results: {'eval_loss': 0.29126086831092834, 'eval_model_preparation_time': 0.0032, 'eval_precision': 0.825381679389313, 'eval_recall': 0.9010416666666666, 'eval_f1': 0.8615537848605578, 'eval_accuracy': 0.9841509742249092, 'eval_runtime': 2.3923, 'eval_samples_per_second': 393.338, 'eval_steps_per_second': 49.324, 'epoch': 8.0}


With the fixed metrics function, you should see much better results. Let's examine these final evaluation metrics:

| Metric | Value | Explanation |
|--------|-------|-------------|
| eval_loss | 0.2913 | The final loss value on the test dataset. Lower values indicate better model fit. |
| eval_precision | 0.8254 | The proportion of predicted disease entities that were actually correct. Our model is right about 82.5% of the time when it identifies a disease. |
| eval_recall | 0.9010 | The proportion of actual disease entities that were correctly identified. Our model finds about 90% of all diseases in the text. |
| eval_f1 | 0.8616 | The harmonic mean of precision and recall, providing a single metric to judge overall performance. This strong F1 score of 0.86 indicates excellent performance for a biomedical NER task. |
| eval_accuracy | 0.9842 | The proportion of all tokens (including non-entities) that were correctly classified. The high value reflects both good entity detection and correct identification of non-entities. |
| eval_runtime | 2.3923 | Time in seconds to evaluate the model on the test set. |
| eval_samples_per_second | 393.34 | The number of test examples processed per second. |

These results demonstrate that our model has successfully learned to identify disease mentions in medical text with high accuracy. The high recall (90%) is particularly valuable in clinical applications where missing a disease mention could have more serious consequences than occasionally mislabeling a non-disease term.

# 6. Qualitative Examples: Before vs. After Fine-Tuning
Numbers are important, but it's also useful to see the model’s predictions on actual text to understand what it’s doing. We’ll take a synthetic clinical note snippet and run the NER model on it. We'll compare the output before and after fine-tuning to appreciate the improvement.

Let's create a pipeline for NER using our fine-tuned model:

In [76]:
# Run predictions on a few examples from the test set
from transformers import pipeline

ner_pipeline = pipeline("ner", model=model, tokenizer=tokenizer, aggregation_strategy="simple")

# Test on a few examples from the test set
for i in range(3):  # Check first 3 examples
    example = raw_datasets["test"][i]
    text = " ".join(example["tokens"])
    print(f"\nExample {i+1} text: {text}")
    print("True entities:", [(example["tokens"][j], id2label[tag]) for j, tag in enumerate(example["ner_tags"]) if tag != 0])

    # Get model predictions
    entities = ner_pipeline(text)
    print("Predicted entities:", [(ent['word'], ent['entity_group']) for ent in entities])


Example 1 text: Clustering of missense mutations in the ataxia - telangiectasia gene in a sporadic T - cell leukaemia .
True entities: [('ataxia', 'B-Disease'), ('-', 'I-Disease'), ('telangiectasia', 'I-Disease'), ('sporadic', 'B-Disease'), ('T', 'I-Disease'), ('-', 'I-Disease'), ('cell', 'I-Disease'), ('leukaemia', 'I-Disease')]
Predicted entities: [('ataxia - telangiectasia', 'Disease'), ('sporadic', 'Disease'), ('t - cell leukaemia', 'Disease')]

Example 2 text: Ataxia - telangiectasia ( A - T ) is a recessive multi - system disorder caused by mutations in the ATM gene at 11q22 - q23 ( ref . 3 ) .
True entities: [('Ataxia', 'B-Disease'), ('-', 'I-Disease'), ('telangiectasia', 'I-Disease'), ('A', 'B-Disease'), ('-', 'I-Disease'), ('T', 'I-Disease'), ('recessive', 'B-Disease'), ('multi', 'I-Disease'), ('-', 'I-Disease'), ('system', 'I-Disease'), ('disorder', 'I-Disease')]
Predicted entities: [('at', 'Disease'), ('##axia - telangiectasia', 'Disease'), ('a - t', 'Disease'), ('recessive

By setting aggregation_strategy="simple", the pipeline will group contiguous tokens with the same entity tag, so we get whole entity mentions rather than token-by-token output. Now, consider this sample text (simulating a line from a clinical note):

The fine-tuned model correctly detects “diabetes mellitus” and “hypertension” as medical problems (diseases), with high confidence scores. This is a big improvement from before training, when those terms would not be identified at all.

*Discussion*: Our model treats both diabetes and hypertension as Disease entities because the dataset we trained on is focused on disease names​. In a real clinical NER scenario, "medical problems" could include symptoms and other conditions as well. If we had a dataset that labeled symptoms, we could fine-tune in a similar way. The key takeaway is that with domain-specific data and a pretrained model, we can adapt a Transformer to accurately tag clinically relevant information.

# 7. Conclusion and Next Steps
In this tutorial, we covered the end-to-end process of fine-tuning a Transformer model (BioClinicalBERT) to perform NER on clinical text. Starting from an open dataset​
huggingface.co
, we preprocessed the data, loaded a domain-specific model​
huggingface.co
, and trained it to recognize disease entities. We evaluated the model’s performance, seeing a drastic improvement in precision/recall/F1 from the untrained baseline to the fine-tuned model. We also tested the model on example text to see how it identifies medical problems in context.

**Key learnings and tips:**
- Data preprocessing: Aligning labels with tokenized inputs is essential for token classification tasks. Libraries like 🤗 Datasets and fast tokenizers make this easier.
- Choosing a model: A model pretrained on biomedical and clinical text can significantly boost performance for medical NER compared to a general-language model​.
- Training: Even a few epochs of fine-tuning with a reasonably low learning rate (2e-5) can achieve high accuracy for NER when starting from a good pretrained base.
- Evaluation: Always look at precision, recall, and F1 in NER tasks – token-level accuracy can be misleading when the dataset has many non-entity tokens.
- Qualitative checks: It’s helpful to try the model on realistic examples to verify it’s picking up the entities of interest (and to ensure it’s not flagging spurious ones).


With this foundation, you can experiment further. For example, you could try a different dataset (e.g., one that includes symptoms or other entity types), or use a larger model like BioMegatron or a Clinical-XLM-R for multilingual clinical notes. You could also incorporate more entity categories (like medications, tests, treatments) if your dataset provides them.

By understanding and following these steps, healthcare professionals and researchers can train custom NER models to extract valuable information from clinical narratives, which can be a stepping stone to building clinical NLP applications such as automated problem list generation, clinical decision support, or research data mining.